File: problems/29-advanced-trie/002-maximum-xor-pair/java/Solution.java
public class Solution {
    private static class Node {
        Node[] child = new Node[2];
    }

    private final Node root = new Node();

    private void insert(int value) {
        Node current = root;

        for (int bit = 30; bit >= 0; bit--) {
            int currentBit = (value >> bit) & 1;
            if (current.child[currentBit] == null) {
                current.child[currentBit] = new Node();
            }
            current = current.child[currentBit];
        }
    }

    private int bestXor(int value) {
        Node current = root;
        int result = 0;

        for (int bit = 30; bit >= 0; bit--) {
            int currentBit = (value >> bit) & 1;
            int wantedBit = currentBit ^ 1;

            if (current.child[wantedBit] != null) {
                result |= (1 << bit);
                current = current.child[wantedBit];
            } else {
                current = current.child[currentBit];
            }
        }

        return result;
    }

    public int findMaximumXOR(int[] nums) {
        if (nums.length < 2) {
            return 0;
        }

        int answer = 0;
        insert(nums[0]);

        for (int i = 1; i < nums.length; i++) {
            answer = Math.max(answer, bestXor(nums[i]));
            insert(nums[i]);
        }

        return answer;
    }
}

File: problems/29-advanced-trie/002-maximum-xor-pair/python/solution.py
class Solution:
    def find_maximum_xor(self, nums: list[int]) -> int:
        if len(nums) < 2:
            return 0

        root: dict[int, dict] = {}

        def insert(value: int) -> None:
            node = root
            for bit in range(30, -1, -1):
                current_bit = (value >> bit) & 1
                node = node.setdefault(current_bit, {})

        def best_xor(value: int) -> int:
            node = root
            result = 0

            for bit in range(30, -1, -1):
                current_bit = (value >> bit) & 1
                wanted_bit = current_bit ^ 1

                if wanted_bit in node:
                    result |= 1 << bit
                    node = node[wanted_bit]
                else:
                    node = node[current_bit]

            return result

        answer = 0
        insert(nums[0])

        for value in nums[1:]:
            answer = max(answer, best_xor(value))
            insert(value)

        return answer
